Scripts: Add CSS support to start and build scripts (#21730)
WordPress/gutenberg: packages/scripts/README.md

* [Entry](https://webpack.js.org/configuration/entry-context/#entry): `src/index.js`
* [Output](https://webpack.js.org/configuration/output): `build/index.js`
* [Loaders](https://webpack.js.org/loaders/):
  * [`babel-loader`](https://webpack.js.org/loaders/babel-loader/) allows transpiling JavaScript files using Babel and webpack.
  * [`@svgr/webpack`](https://www.npmjs.com/package/@svgr/webpack) and [`url-loader`](https://webpack.js.org/loaders/url-loader/) makes it possible to handle SVG files in JavaScript code.

>   * [`css-loader`](https://webpack.js.org/loaders/css-loader/) chained with [`postcss-loader`](https://webpack.js.org/loaders/postcss-loader/) and [sass-loader](https://webpack.js.org/loaders/sass-loader/) let webpack processing CSS, SASS or SCSS files referenced in JavaScript files and extract into CSS files with [`MiniCssExtractPlugin`](https://webpack.js.org/plugins/mini-css-extract-plugin/).

Reviewer: ```suggestion
  * [`css-loader`](https://webpack.js.org/loaders/css-loader/) chained with [`postcss-loader`](https://webpack.js.org/loaders/postcss-loader/) and [sass-loader](https://webpack.js.org/loaders/sass-loader/) let webpack process CSS, SASS or SCSS files referenced in JavaScript files.
```
This sentence feels long, and `MiniCssExtractPlugin` is mentioned immediately in the next section

* [Plugins](https://webpack.js.org/configuration/plugins) (among others):
  * [`MiniCssExtractPlugin`](https://webpack.js.org/plugins/mini-css-extract-plugin/) extracts CSS into separate files. It creates a CSS file per JavaScript entry point which contains CSS.
  * [`@wordpress/dependency-extraction-webpack-plugin`](/packages/dependency-extraction-webpack-plugin/README.md) is used with the default configuration to ensure that WordPress provided scripts are not included in the built bundle.

#### Using CSS

_Example:_

```scss
// index.scss
$body-color: red;

.wp-block-my-block {
    color: $body-color;
}
```

```css
/* style.css */
.wp-block-my-block {
    background-color: black;
}
```

```js
// index.js
import './index.scss';
import './style.css';
```

> When you run the build using the default command `wp-scripts build` (also applies to `start`) in addition to JavaScript file `index.js` generated in the `dist` folder, you should see two more files:

Reviewer: ```suggestion
When you run the build using the default command `wp-scripts build` (also applies to `start`) in addition to the JavaScript file `index.js` generated in the `build` folder, you should see two more files:
```
Default output is `build`.

> 1. `index.css` – in general, the majority of JavaScript code written for the block, it's going to be rendered in the block editor. Therefore all imported files are bundled into one chunk named after the entry point that defaults to `index.js`, and thus the file created becomes `index.css`.

Reviewer: I think this is more technical than it needs to be, if someone's just looking for CSS info. Maybe:

```suggestion
1. `index.css` – All imported CSS files are bundled into one chunk named after the entry point, which defaults to `index.js`, and thus the file created becomes `index.css`. This is for styles used only in the editor.
```

2. `style.css` – imported `style.css` file(s) (applies to SASS and SCSS extensions) get bundled into one `style.css` file that is meant to be used both on the front-end and in the editor.

You can also have multiple entry points as described in the docs for the script:
```bash
wp-scripts start entry-one.js entry-two.js --output-path=custom
```

If you do so, then CSS files generated will follow the names of the entry points: `entry-one.css` and `entry-two.css`.

> You can have only one entry point and name it differently if you will. The chances that you would call an entry point `style.js` is close to zero, but possible and therefore, it needs to be discouraged.

Reviewer: ```suggestion
You can have only one entry point and name it differently if you will. Avoid using `style` for an entry point name, this will break your build process.
```

* [Plugins](https://webpack.js.org/configuration/plugins) (among others):
  * [`MiniCssExtractPlugin`](https://webpack.js.org/plugins/mini-css-extract-plugin/) extracts CSS into separate files. It creates a CSS file per JavaScript entry point which contains CSS.
  * [`@wordpress/dependency-extraction-webpack-plugin`](/packages/dependency-extraction-webpack-plugin/README.md) is used with the default configuration to ensure that WordPress provided scripts are not included in the built bundle.

#### Using CSS

_Example:_

```scss
// index.scss
$body-color: red;

.wp-block-my-block {
    color: $body-color;
}
```

```css
/* style.css */
.wp-block-my-block {
    background-color: black;
}
```

```js
// index.js
import './index.scss';
import './style.css';
```

When you run the build using the default command `wp-scripts build` (also applies to `start`) in addition to the JavaScript file `index.js` generated in the `build` folder, you should see two more files:
1. `index.css` – all imported CSS files are bundled into one chunk named after the entry point, which defaults to `index.js`, and thus the file created becomes `index.css`. This is for styles used only in the editor.

> 2. `style.css` – imported `style.css` file(s) (applies to SASS and SCSS extensions) get bundled into one `style.css` file that is meant to be used both on the front-end and in the editor.

Reviewer 2: Have we previously made such assumptions about how `@wordpress/scripts` is used for bundling for the editor, vs. being purely general-purpose in its goals? I need to spend some time to look at some of the history here.

Author: It's something I proposed based on the previous implementation from https://github.com/WordPress/gutenberg/pull/14847, an article from @phpbits https://jeffreycarandang.com/create-gutenberg-block-plugin-wp-scripts-postcss-build/, `create-guten-block` (https://github.com/ahmadawais/create-guten-block/blob/bdc40192fc7fa30519d821e2aa835b2aa6ef7534/packages/cgb-scripts/config/webpack.config.prod.js#L109-L118) or the old implementation in Gutenberg before we moved to packages. I made an assumption that it became a standard approach in the community.

To extend the provided webpack config, or replace subsections within the provided webpack config, you can provide your own `webpack.config.js` file, `require` the provided `webpack.config.js` file, and use the [`spread` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) to import all of or part of the provided configuration.

> In the example below, a `webpack.config.js` file is added to the root folder extending the provided webpack config to include custom logic to parse module's source and convert it to a JavaScript object using [`toml`](https://www.npmjs.com/package/toml). It may be useful to import toml or other non-JSON files as JSON, without specific loaders:

Reviewer 2: Is the idea with this change that we now build-in CSS support, so we need a different example of extending? If so, shouldn't we also have something here in the documentation which describes the new CSS support?

Author: Yes, this example won’t work anymore so I found something else. 

Yes, I mentioned that documentation is missing in https://github.com/WordPress/gutenberg/pull/21730#issuecomment-621326806.
